core/sr-primitives/src/sr_arithmetic.rs

```diff
@@ -0,0 +1,1013 @@
+
+			/// Everything.
+			pub fn one() -> Self { Self($max) }
+
+			/// consume self and deconstruct into a raw numeric type.
+			pub fn deconstruct(self) -> $type { self.0 }
+
+			/// return the scale at which this per-thing is working.
+			pub const fn accuracy() -> $type { $max }
+
+			/// From an explicitly defined number of parts per maximum of the type.
+			///
+			/// This can be called at compile time.
+			pub const fn from_parts(parts: $type) -> Self {
+				Self([parts, $max][(parts > $max) as usize])
+			}
+
+			/// Converts from a percent. Equal to `x / 100`.
+			///
+			/// This can be created at compile time.
+			pub const fn from_percent(x: $type) -> Self {
+				Self([x, 100][(x > 100) as usize] * ($max / 100))
+			}
+
+			/// Return the product of multiplication of this value by itself.
+			pub fn square(self) -> Self {
+				// both can be safely casted and multiplied.
+				let p: $upper_type = self.0 as $upper_type * self.0 as $upper_type;
+				let q: $upper_type = $max as $upper_type * $max as $upper_type;
+				Self::from_rational_approximation(p, q)
+			}
+
+			/// Converts a fraction into `Permill`.
+			#[cfg(feature = "std")]
+			pub fn from_fraction(x: f64) -> Self { Self((x * ($max as f64)) as $type) }
+
+			/// Approximate the fraction `p/q` into a per-thing fraction. This will never overflow.
+			///
+			/// The computation of this approximation is performed in the generic type `N`. Given
+			/// `M` as the data type that can hold the maximum value of this per-thing (e.g. u32 for
+			/// perbill), this can only work if `N == M` or `N: From<M> + TryInto<M>`.
+			pub fn from_rational_approximation<N>(p: N, q: N) -> Self
+				where N: Clone + Ord + From<$type> + TryInto<$type> + ops::Div<N, Output=N>
+			{
+			// TODO: double check with @thiolliere the new bound. If your fraction type has a max
+			// that fits in u64, then you cannot build a rational from it that is `2u8 / 4u8` So
+			// using `SimpleArithmetic` here was okay as long as we don't add a new fraction type
+			// which is embedded in a u64. With perquintill we have to change this.
+				// q cannot be zero.
+				let q = q.max((1 as $type).into());
+				// p should not be bigger than q.
+				let p = p.min(q.clone());
+
+				let factor = (q.clone() / $max.into()).max((1 as $type).into());
+
+				// q cannot overflow: (q / (q/$max)) < $max. p < q hence p also cannot overflow.
+				let p_reduce: $type = (p / factor.clone()).try_into().unwrap_or_else(|_| panic!());
+				let q_reduce: $type = (q / factor.clone()).try_into().unwrap_or_else(|_| panic!());
+				// `p_reduced` and `q_reduced` are withing $type. Mul by another $max will always
+				// fit in $upper_type. This is guaranteed by the macro tests.
+				let part =
+					p_reduce as $upper_type
+					* ($max as $upper_type)
+					/ q_reduce as $upper_type;
+
+				$name(part as $type)
+			}
+		}
+
+		impl Saturating for $name {
+			fn saturating_add(self, rhs: Self) -> Self {
+				Self::from_parts(self.0 + rhs.0)
```

Comment:
this is also guaranteed not to overflow because of tests no ? like imagine we want to implement PerTenquintill then this operation should have been done with saturating_add.

Actually maybe we can do saturating_add here just so its sure anyway. (even if it is already sure for our types)

Reply:
Now with the new assertion that `$max < $type::max_value()` this is safe right? `self.0` and `rhs.0` are both less than `$max` and will still fit into `$type`, and the `from_parts` coerces it to `$max` if it is more than `$max`. 

will make it saturating anyhow.

Reply:
yes ok I'm ok as is actually